### PR TITLE
Fix right click in selection of additional caret

### LINKED ISSUE #76417: Inconsistant behavior when right clicking in a multiple selection
### Godot version

4.0.2

### System information

Windows 10, Linux

### Issue description

# Issue : 
When having multiple selection, right clicking in one of them didn't produce always the same effect.

Right click in the selection of the main caret will disabled all other selections and act like if it was only this selection.

Right click in all other selection, will disable all selection and show the popup that normaly pops when there isn't any selection.

Seems that multicarret is not correctly managed here : it's only check if mouse is in selection for the main caret and considere that mouse is outside any selection as long as it's outside the first caret selection.

Notice that @Paulb23 report me that for him, no selection is removed. So this issue may need testing for bug confirmation or not.

mutlicaret

# Expected behaviour

Whatever caret the selection belongs to, it should keep only this selection and show the popup for this selection.

### Steps to reproduce

Select several parts of code maintaining alt key to do a multiple selection.

Right click on the first one : other selection disapeared (as expected I think), selection menu popup.

Do it again but click on the second selection (not the main caret selection) : all selection are discared and no selection menu pops.

### Minimal reproduction project

Any script will do the job.

### PR DESCRIPTION
**Pull request by jmb462**
Fix #76417

# Issue description

Right click in the selection of any secondary caret discard all selections and show context menu has is there was no selection.

# Investigations

In ``ScriptTextEditor::_text_edit_gui_input(const Ref<InputEvent> &ev)`` multicaret is not managed when checking if click was in a selection. It only check the main caret selection. If it's outside the main caret selection, it discards all secondary caret and all selections.

# Fix proposal
(modified version after taking comments into account)

Managed multicarret in ``ScriptTextEditor::_text_edit_gui_input(const Ref<InputEvent> &ev)`` : iterate throw each caret selection to check if right click is in any of the caret selections. Keep all selection.

Same modification in ``void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input)`` : only one caret selection was checked. Iterates throw all selections and keep it all.

# Before

mutlicaret

# After

fixed

### CODE REVIEW

**Comment by ajreckof** [comment]
Is there a particular reason we should only keep one of the selection and not all of them?

**Comment by jmb462** [comment]
> Is there a particular reason we should only keep one of the selection and not all of them?

In VS Code, we can copy multiselection and selections are each pasted on a different line.
So maybe we could also keep all selections and do the same.

**Comment by ajreckof** [comment]
> > Is there a particular reason we should only keep one of the selection and not all of them?
> 
> In VS Code, we can copy multiselection and selections are each pasted on a different line. So maybe we could also keep all selections and do the same.

Your coment made me realise that it is already the case for keyboard shortcut and therefore it would be even stranger not to do it for right click.

**Comment by jmb462** [comment]
Turning this PR into a draft.

Thanks to @ajreckof comments, it clearly appears that we have to quit all carets when right clicking on a selection.
I will modify this PR in this way (almost done).

File: scene/gui/text_edit.cpp [hunk 1]
```diff
@@ -1851,23 +1851,29 @@ void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
...
+					if (has_selection()) {
+						for (int i = 0; i < get_caret_count(); i++) {
+							int from_line = get_selection_from_line(i);
+							int to_line = get_selection_to_line(i);
+							int from_column = get_selection_from_column(i);
+							int to_column = get_selection_to_column(i);
+
+							if (row >= from_line && row <= to_line && (row != from_line || col >= from_column) && (row != to_line || col <= to_column)) {
+								// Right click in one of the selected text
+								selection_clicked = true;
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp:1866]: We should be able to `break` early here.

File: scene/gui/text_edit.cpp [hunk 2]
```diff
@@ -1851,23 +1851,29 @@ void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
...
+							}
+						}
+						if (!selection_clicked) {
+							deselect();
 						}
 					}
-					if (!has_selection(caret)) {
-						set_caret_line(row, true, false, 0, caret);
-						set_caret_column(col, true, caret);
+					if (!has_selection()) {
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp:1873]: Could we check if `selection_clicked` is `false` here? would be able to move the `deselect()` call here as well.

File: editor/plugins/script_text_editor.cpp [hunk 3]
```diff
@@ -1805,31 +1805,38 @@ void ScriptTextEditor::_text_edit_gui_input(const Ref<InputEvent> &ev) {
...
-					// Right click is outside the selected text
+				for (int i = 0; i < tx->get_caret_count(); i++) {
+					int from_line = tx->get_selection_from_line(i);
+					int to_line = tx->get_selection_to_line(i);
+					int from_column = tx->get_selection_from_column(i);
+					int to_column = tx->get_selection_to_column(i);
+
+					if (row >= from_line && row <= to_line && (row != from_line || col >= from_column) && (row != to_line || col <= to_column)) {
+						// Right click in one of the selected text
+						caret_clicked = i;
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp:1819]: Same here as below.

**Review by Paulb23 — COMMENTED** [review]
Nice find and fix! 

Couple small notes inline

File: scene/gui/text_edit.cpp (hunk 2 above)
Comment (jmb462) [reply scene/gui/text_edit.cpp]: Thanks, done.

**Review by Paulb23 — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 4.0.3.

**Pull request merged**